clients/java/signalr/src/main/java/com/microsoft/aspnet/signalr/HubConnection.java

```diff
@@ -151,79 +156,97 @@ public HubConnectionState getConnectionState() {
+        negotiateResponse = Negotiate.processNegotiate(url, accessToken);
+
+        if (negotiateResponse.getConnectionId() != null) {
+            if (url.contains("?")) {
+                url = url + "&id=" + negotiateResponse.getConnectionId();
+            } else {
+                url = url + "?id=" + negotiateResponse.getConnectionId();
             }
         }
 
-        logger.log(LogLevel.Debug, "Starting HubConnection");
-        if (transport == null) {
-            transport = new WebSocketTransport(url, logger, headers);
+        if (negotiateResponse.getAccessToken() != null) {
+            this.headers.put("Authorization", "Bearer " + negotiateResponse.getAccessToken());
+        }
+
+        if (negotiateResponse.getRedirectUrl() != null) {
+            url = this.negotiateResponse.getRedirectUrl();
         }
 
-        transport.setOnReceive(this.callback);
-        transport.start();
-        String handshake = HandshakeProtocol.createHandshakeRequestMessage(new HandshakeRequestMessage(protocol.getName(), protocol.getVersion()));
-        transport.send(handshake);
-        connectionState = HubConnectionState.CONNECTED;
-        logger.log(LogLevel.Information, "HubConnected started.");
+        return negotiateResponse;
     }
 
     /**
      * Stops a connection to the server.
      */
     private void stop(String errorMessage) {
-        if (connectionState == HubConnectionState.DISCONNECTED) {
-            return;
-        }
+        hubConnectionStateLock.lock();
+        try {
 
-        if (errorMessage != null) {
-            logger.log(LogLevel.Error, "HubConnection disconnected with an error %s.", errorMessage);
-        } else {
-            logger.log(LogLevel.Debug, "Stopping HubConnection.");
-        }
 
-        transport.stop();
-        connectionState = HubConnectionState.DISCONNECTED;
-        logger.log(LogLevel.Information, "HubConnection stopped.");
-        if (onClosedCallbackList != null) {
-            HubException hubException = new HubException(errorMessage);
-            for (Consumer<Exception> callback : onClosedCallbackList) {
-                callback.accept(hubException);
+            if (connectionState == HubConnectionState.DISCONNECTED) {
+                return;
+            }
+
+            if (errorMessage != null) {
+                logger.log(LogLevel.Error, "HubConnection disconnected with an error %s.", errorMessage);
+            } else {
+                logger.log(LogLevel.Debug, "Stopping HubConnection.");
+            }
+
+            transport.stop();
+            connectionState = HubConnectionState.DISCONNECTED;
+            logger.log(LogLevel.Information, "HubConnection stopped.");
+            if (onClosedCallbackList != null) {
+                HubException hubException = new HubException(errorMessage);
+                for (Consumer<Exception> callback : onClosedCallbackList) {
```

Comment:
This looks like good deadlock potential. It is common practice to try and restart the connection in the `onClosed` callbacks which are currently run inside this lock. And that would mean calling `start`, which trys to acquire the lock resulting in a deadlock.

Reply:
Yeah, good eye. We just need to reduce the scope of the locked operations.